### PR TITLE
8255670: Improve C2's detection of modified nodes

### PR DESCRIPTION
**Pull request by eme64**
Added `record_modified_node` to:
```
Node::clone
Node::add_req
Node::add_req_batch
Node::ins_req
Node::add_prec
Node::rm_prec
Node::set_prec
```

Added `igvn->_worklist.push(node)` in various places that modified a `node` but did not add it to the igvn worklist.

7 times I had to push `Root`, 5 of these it was because of the creation of a `HaltNode`, which means we have a `root->add_req(halt)`. 

In one case we have a MergeMemStream node, which gets two MergeMem nodes as input, and streams over them.
Unfortunately, it modifies one of the two, which then can trigger our assertion code. I now push this node to the igvn worklist, but a better fix would be to make MergeMemStream leave the MergeMem nodes unmodified. I think that should be possible, filed an RFE [JDK-8293358](https://bugs.openjdk.org/browse/JDK-8293358)

Other small refactorings (review suggestions):
- Refactored a few cases of `igvn.worklist.push` and `set_req` with `set_req_X`.
- Removed unecessary `igvn.worklist.push` before `replace_input_of`.
- Refactored some code with `delete_precedence_of`, `add_input_to` and `replace_input_of`.

FYI:
What I am NOT doing here, and leave to a future RFE/independent change: investigate / implement these assertions for late/incremental inlining [JDK-8293362](https://bugs.openjdk.org/browse/JDK-8293362).

Ran larger regression tests, and 7-9h of fuzzing on 3 platforms.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8255670](https://bugs.openjdk.org/browse/JDK-8255670): Improve C2's detection of modified nodes

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [84091bd8](https://git.openjdk.org/jdk/pull/9439/files/84091bd8d5e3250ed1dcced9335d74471fa289a3)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [84091bd8](https://git.openjdk.org/jdk/pull/9439/files/84091bd8d5e3250ed1dcced9335d74471fa289a3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9439/head:pull/9439` \
`$ git checkout pull/9439`

Update a local copy of the PR: \
`$ git checkout pull/9439` \
`$ git pull https://git.openjdk.org/jdk pull/9439/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9439`

View PR using the GUI difftool: \
`$ git pr show -t 9439`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9439.diff">https://git.openjdk.org/jdk/pull/9439.diff</a>

</details>

### CODE REVIEW

**Comment by eme64** [comment]
I'm back to work. Will spend time on this during the next weeks.

File: src/hotspot/share/opto/loopnode.cpp [hunk 1]
```diff
@@ -5126,6 +5132,7 @@ int PhaseIdealLoop::build_loop_tree_impl( Node *n, int pre_order ) {
           _igvn.register_new_node_with_optimizer(halt);
           set_loop(halt, l);
           C->root()->add_req(halt);
+          _igvn._worklist.push(C->root());
```
Comment (rwestrel) [inline src/hotspot/share/opto/loopnode.cpp:5135]: Maybe add a method to PhaseIterGVN that does the add_req + push similar to replace_input_of?

**Review by vnkozlov — COMMENTED** [review]
The only GVN transformation we do with Root node is removing TOP inputs in `RootNode::Ideal()`. That is why we are "sloppy" about putting it on worklist when we know that added input is not TOP (new Halt node in this case).

**Comment by eme64** [comment]
> The only GVN transformation we do with Root node is removing TOP inputs in `RootNode::Ideal()`. That is why we are "sloppy" about putting it on worklist when we know that added input is not TOP (new Halt node in this case).

@vnkozlov right, not adding Root to the worklist was not a bug in these cases. Being "sloppy" in these cases was ok.
However, the question is how we want to handle it now. Because `modified_node` has Root recorded after the `add_req`, so we need to either add Root to the worklist, or take it off the `modified_node`.
I discussed it with @TobiHartmann , and we think that adding it to the worklist is probably the clean solution. If any node has a modification, it is possible that IGVN needs to optimize it, or at least that there could be a future changeset that adds such an optimization.
Alternatives: remove Root from `modified_node` after `add_req` (Ad-Hoc solution), or completely exclude Root from the assert.
@vnkozlov What do you think?

File: src/hotspot/share/opto/loopnode.cpp (hunk 1 above)
Comment (eme64) [reply src/hotspot/share/opto/loopnode.cpp:5135]: @rwestrel Thanks for the suggestion, I will do that

**Comment by vnkozlov** [comment]
> > The only GVN transformation we do with Root node is removing TOP inputs in `RootNode::Ideal()`. That is why we are "sloppy" about putting it on worklist when we know that added input is not TOP (new Halt node in this case).
> 
> @vnkozlov right, not adding Root to the worklist was not a bug in these cases. Being "sloppy" in these cases was ok. However, the question is how we want to handle it now. Because `modified_node` has Root recorded after the `add_req`, so we need to either add Root to the worklist, or take it off the `modified_node`. I discussed it with @TobiHartmann , and we think that adding it to the worklist is probably the clean solution. If any node has a modification, it is possible that IGVN needs to optimize it, or at least that there could be a future changeset that adds such an optimization. Alternatives: remove Root from `modified_node` after `add_req` (Ad-Hoc solution), or completely exclude Root from the assert. @vnkozlov What do you think?

Okay, it make sense. I agree with changes.

File: src/hotspot/share/opto/loopnode.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/loopnode.cpp:5135]: Yes, it is good suggestion.

**Comment by eme64** [comment]
I implemented the requested changes, and performed full testing again, including fuzzing.
@rwestrel @vnkozlov Can you please review the PR again?

**Review by vnkozlov — APPROVED** [review]
Nice.

File: src/hotspot/share/opto/phaseX.hpp [hunk 2]
```diff
@@ -531,12 +531,26 @@ class PhaseIterGVN : public PhaseGVN {
...
   }
 
+  // Delete precedence edge i of "n"
+  void delete_precedence_of(Node* n, int i) {
+    rehash_node_delayed(n);
+    n->rm_prec(i);
+  }
+
+
+
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/phaseX.hpp:553]: Please remove two of the newlines.

**Review by TobiHartmann — APPROVED** [review]
Looks good!

File: src/hotspot/share/opto/phaseX.hpp (hunk 2 above)
Comment (eme64) [reply src/hotspot/share/opto/phaseX.hpp]: thanks, done

**Review by rwestrel — APPROVED** [review]
Looks good to me. Thanks for making the changes.

**Review by TobiHartmann — APPROVED** [review]

**Comment by eme64** [comment]
Thanks @TobiHartmann @rwestrel @vnkozlov for the help with understanding the fix, and of course the reviews!
/integrate

**Pull request closed without merging**